autoload/ale/go.vim

```diff
@@ -0,0 +1,23 @@
+" Author: Horacio Sanson https://github.com/hsanson
+" Description: Functions for integrating with Go tools
+
+" Find the nearest dir listed in GOPATH and assume it the root of the go
+" project.
+function! ale#go#FindProjectRoot(buffer) abort
+
+    let l:sep = has('win32') ? ';' : ':'
+
+    let l:filename = ale#path#Simplify(expand('#' . a:buffer . ':p'))
+
+    for l:name in split($GOPATH, l:sep)
+      let l:path_dir = ale#path#Simplify(l:name)
+
+      " Use the directory from GOPATH if the current filename starts with it.
+      if l:filename[: len(l:path_dir) - 1] is? l:path_dir
+          return l:path_dir
+      endif
+    endfor
+
+    return ''
```

Comment:
Note this function won't work with go modules. It should return any directory with a `go.mod` file as a root directory too.

Reply:
Actually, what is this intended to return? If the code is in `/home/martin/go/src/github.com/me/project/foo/code.go`, then should it return `/home/martin/go/src/github.com/me/project` or `/home/martin/go`? The function name suggests the former, but it actually seems to do the latter.

Reply:
@Carpetsmoker I have tested with both root paths and go-langserver works perfectly. Even setting the root path to /tmp or $USER works ok. I can get auto-completion for files in any project installed in my machine. 

So I opted to set as root the first path that matches an entry in $GOPATH.